x/cronos/keeper/precompiles/base_contract.go

```diff
@@ -10,18 +12,56 @@ type Registrable interface {
 
 type BaseContract interface {
 	Registrable
+	RequiredGas(input []byte) uint64
 }
 
 type baseContract struct {
-	address common.Address
+	address                       common.Address
+	kvGasConfig                   storetypes.GasConfig
+	nameByMethod                  map[[4]byte]string
+	gasByMethod                   map[[4]byte]uint64
+	emptyGasIfInputLessThanPrefix bool
+	logger                        log.Logger
 }
 
-func NewBaseContract(address common.Address) BaseContract {
+func NewBaseContract(
+	address common.Address,
+	kvGasConfig storetypes.GasConfig,
+	nameByMethod map[[4]byte]string,
+	gasByMethod map[[4]byte]uint64,
+	emptyGasIfInputLessThanPrefix bool,
+	logger log.Logger,
+) BaseContract {
 	return &baseContract{
-		address: address,
+		address,
+		kvGasConfig,
+		nameByMethod,
+		gasByMethod,
+		emptyGasIfInputLessThanPrefix,
+		logger,
 	}
 }
 
 func (c *baseContract) RegistryKey() common.Address {
 	return c.address
 }
+
+// RequiredGas calculates the contract gas use
+func (c *baseContract) RequiredGas(input []byte) (gas uint64) {
+	var methodID [4]byte
+	copy(methodID[:], input[:4])
+	inputLen := len(input)
+	defer func() {
+		method := c.nameByMethod[methodID]
+		c.logger.Info("required", "gas", gas, "method", method, "len", inputLen)
```

Comment:
I think what we need to log is the actual gas consumed during the method execution to compare with the value returned by the method.

The goal is to be able to collect data more easily to tune the "hard-coded" value in the map later on.


Another idea would be to self adjust the gas during the execution (instead of using the hard coded map) but this could lead to some performance degradation

Reply:
you mean call estimate gas from [relayer](https://github.com/crypto-org-chain/relayer/pull/6/files)?

Reply:
I was more thinking about getting the gas consumed before and after the execution there
https://github.com/crypto-org-chain/cronos/blob/main/x/cronos/keeper/precompiles/bank.go#L132

Reply:
cosmos tx gas can be trace from relayer without precompile config, the gas meter consumed is same as the log from [RequiredGas](https://github.com/crypto-org-chain/cronos/pull/1223/files#diff-b93d562e3e27735b0f3311c3b6cf9b278f98f5a1fe03b09cfe232ef752365497R100)